File: nine-language-version/lessons/03-control-flow/control_flow.rs
#!/usr/bin/env rust-script
/*
 * Lesson 3: Control Flow in Rust
 * Demonstrates conditionals, loops, pattern matching, and expressions
 */

fn main() {
    println!("=== Rust Control Flow ===\n");

    // 1. Basic conditionals (if is an expression)
    println!("1. Conditionals (if expressions):");
    let age = 20;

    // If as statement
    if age >= 18 {
        println!("  Age {}: Adult", age);
    } else if age >= 13 {
        println!("  Age {}: Teenager", age);
    } else {
        println!("  Age {}: Child", age);
    }

    // If as expression
    let status = if age >= 18 { "Adult" } else { "Minor" };
    println!("  Status: {}", status);

    // 2. For loops
    println!("\n2. For Loops:");
    println!("  Count to 5:");
    print!("   ");
    for i in 0..5 {
        print!(" {}", i);
    }
    println!();

    println!("  Iterate array:");
    let fruits = vec!["apple", "banana", "cherry"];
    for fruit in &fruits {
        println!("    {}", fruit);
    }

    println!("  With enumerate:");
    for (i, fruit) in fruits.iter().enumerate() {
        println!("    {}: {}", i, fruit);
    }

    // 3. While loop
    println!("\n3. While Loop:");
    let mut count = 0;
    while count < 3 {
        println!("  Count: {}", count);
        count += 1;
    }

    // 4. Loop (infinite loop with break)
    println!("\n4. Loop (with break):");
    let mut i = 0;
    loop {
        println!("  Iteration: {}", i);
        i += 1;
        if i >= 3 {
            break;
        }
    }

    // Loop as expression (returns value)
    let result = loop {
        i += 1;
        if i >= 10 {
            break i * 2;  // Return value
        }
    };
    println!("  Loop returned: {}", result);

    // 5. Boolean logic
    println!("\n5. Boolean Logic:");
    let x = 5;
    let y = 10;
    println!("  x={}, y={}", x, y);
    println!("  x > 3 && y < 20: {}", x > 3 && y < 20);
    println!("  x > 10 || y > 5: {}", x > 10 || y > 5);
    println!("  !(x == y): {}", !(x == y));

    println!("\n  Only bool type (no truthiness):");
    println!("    Must use explicit boolean expressions");

    // 6. FizzBuzz
    println!("\n6. FizzBuzz (1-20):");
    print!(" ");
    for i in 1..=20 {
        print!(" {}", fizzbuzz(i));
    }
    println!();

    // 7. Match expressions (pattern matching)
    println!("\n7. Match Expressions (pattern matching):");
    println!("  Match on values:");
    for n in 0..=3 {
        println!("    {}: {}", n, describe_number(n));
    }

    println!("\n  Match with guards:");
    for age in &[10, 15, 20] {
        println!("    Age {}: {}", age, describe_age(*age));
    }

    // Match on tuples
    println!("\n  Match on tuples (FizzBuzz style):");
    for i in 1..=5 {
        let result = match (i % 3, i % 5) {
            (0, 0) => "FizzBuzz",
            (0, _) => "Fizz",
            (_, 0) => "Buzz",
            _ => "Number",
        };
        println!("    {}: {}", i, result);
    }

    // 8. If let (pattern matching for single case)
    println!("\n8. If let (single pattern match):");
    let some_value: Option<i32> = Some(42);
    if let Some(x) = some_value {
        println!("  Got value: {}", x);
    } else {
        println!("  Got None");
    }

    let none_value: Option<i32> = None;
    if let Some(x) = none_value {
        println!("  Got value: {}", x);
    } else {
        println!("  Got None");
    }

    // 9. While let (pattern matching loop)
    println!("\n9. While let (pattern matching loop):");
    let mut stack = vec![1, 2, 3];
    while let Some(top) = stack.pop() {
        println!("  Popped: {}", top);
    }

    // 10. Ranges
    println!("\n10. Ranges:");
    println!("  Inclusive range (1..=5):");
    print!("   ");
    for i in 1..=5 {
        print!(" {}", i);
    }
    println!();

    println!("  Exclusive range (1..5):");
    print!("   ");
    for i in 1..5 {
        print!(" {}", i);
    }
    println!();
}

// FizzBuzz using match
fn fizzbuzz(n: i32) -> String {
    match (n % 3, n % 5) {
        (0, 0) => "FizzBuzz".to_string(),
        (0, _) => "Fizz".to_string(),
        (_, 0) => "Buzz".to_string(),
        _ => n.to_string(),
    }
}

// Pattern matching example
fn describe_number(n: i32) -> &'static str {
    match n {
        0 => "Zero",
        1 => "One",
        2 => "Two",
        _ => "Many",
    }
}

// Match with guards
fn describe_age(age: i32) -> &'static str {
    match age {
        n if n >= 18 => "Adult",
        n if n >= 13 => "Teenager",
        _ => "Child",
    }
}

File: nine-language-version/lessons/03-control-flow/control_flow.hs
-- Lesson 3: Control Flow in Haskell
-- Demonstrates conditionals, guards, pattern matching, and recursion

module Main where

main :: IO ()
main = do
    putStrLn "=== Haskell Control Flow ==="
    putStrLn ""

    -- 1. Basic conditionals (if is an expression)
    putStrLn "1. Conditionals (if expressions):"
    let age = 20
    let description = if age >= 18
                      then "Adult"
                      else if age >= 13
                           then "Teenager"
                           else "Child"
    putStrLn $ "  Age " ++ show age ++ ": " ++ description

    -- Ternary-style (one-line if)
    let status = if age >= 18 then "Adult" else "Minor"
    putStrLn $ "  Status: " ++ status

    -- 2. Guards (preferred Haskell style)
    putStrLn "\n2. Guards (idiomatic Haskell):"
    putStrLn $ "  Age " ++ show age ++ ": " ++ describeAge age

    -- 3. No traditional loops - use recursion
    putStrLn "\n3. Recursion (no traditional loops):"
    putStrLn "  Count to 5:"
    putStr "   "
    printNumbers 0 5
    putStrLn ""

    putStrLn "  Iterate list:"
    let fruits = ["apple", "banana", "cherry"]
    mapM_ (\f -> putStrLn $ "    " ++ f) fruits

    -- 4. List comprehensions (alternative to loops)
    putStrLn "\n4. List Comprehensions:"
    putStrLn "  Squares of 1-5:"
    putStr "   "
    putStrLn $ unwords [show (x * x) | x <- [1..5]]

    putStrLn "  Even numbers from 1-10:"
    putStr "   "
    putStrLn $ unwords [show x | x <- [1..10], even x]

    -- 5. Boolean logic
    putStrLn "\n5. Boolean Logic:"
    let x = 5
    let y = 10
    putStrLn $ "  x=" ++ show x ++ ", y=" ++ show y
    putStrLn $ "  x > 3 && y < 20: " ++ show (x > 3 && y < 20)
    putStrLn $ "  x > 10 || y > 5: " ++ show (x > 10 || y > 5)
    putStrLn $ "  not (x == y): " ++ show (not (x == y))

    putStrLn "\n  Only Bool type (no truthiness):"
    putStrLn "    Must use explicit boolean expressions"

    -- 6. FizzBuzz
    putStrLn "\n6. FizzBuzz (1-20):"
    putStr " "
    mapM_ (\x -> putStr $ " " ++ fizzbuzz x) [1..20]
    putStrLn ""

    -- 7. Pattern matching
    putStrLn "\n7. Pattern Matching:"
    putStrLn $ "  describe 0: " ++ describeNumber 0
    putStrLn $ "  describe 1: " ++ describeNumber 1
    putStrLn $ "  describe 5: " ++ describeNumber 5

    -- Pattern matching on lists
    putStrLn "\n  Pattern matching on lists:"
    putStrLn $ "  []: " ++ describeList []
    putStrLn $ "  [1]: " ++ describeList [1]
    putStrLn $ "  [1,2,3]: " ++ describeList [1,2,3]

    -- 8. Case expressions
    putStrLn "\n8. Case Expressions:"
    let day = 3
    let dayType = case day of
          0 -> "Weekend"
          6 -> "Weekend"
          _ | day >= 1 && day <= 5 -> "Weekday"
            | otherwise -> "Invalid day"
    putStrLn $ "  Day " ++ show day ++ ": " ++ dayType

-- Guards example
describeAge :: Int -> String
describeAge age
    | age >= 18 = "Adult"
    | age >= 13 = "Teenager"
    | otherwise = "Child"

-- Recursion example (instead of loops)
printNumbers :: Int -> Int -> IO ()
printNumbers current limit
    | current < limit = do
        putStr $ " " ++ show current
        printNumbers (current + 1) limit
    | otherwise = return ()

-- FizzBuzz with guards
fizzbuzz :: Int -> String
fizzbuzz n
    | n `mod` 15 == 0 = "FizzBuzz"
    | n `mod` 3 == 0  = "Fizz"
    | n `mod` 5 == 0  = "Buzz"
    | otherwise       = show n

-- Pattern matching example
describeNumber :: Int -> String
describeNumber 0 = "Zero"
describeNumber 1 = "One"
describeNumber 2 = "Two"
describeNumber _ = "Many"

-- Pattern matching on lists
describeList :: [Int] -> String
describeList [] = "Empty list"
describeList [_] = "Single element"
describeList [_, _] = "Two elements"
describeList (x:y:_) = "Starts with " ++ show x ++ " and " ++ show y
